apps/prairielearn/src/lib/assets.js

```diff
@@ -83,25 +103,7 @@ module.exports.assetPath = (assetPath) => {
  * @param {string} assetPath - The path to the file inside the `/node_modules` directory.
  */
 module.exports.nodeModulesAssetPath = (assetPath) => {
-  const [maybeScope, maybeModule] = assetPath.split('/');
-  let moduleName;
-  if (maybeScope.indexOf('@') === 0) {
-    // This is a scoped module
-    moduleName = `${maybeScope}/${maybeModule}`;
-  } else {
-    moduleName = maybeScope;
-  }
-
-  // Reading files synchronously and computing cryptographic hashes are both
-  // relatively expensive; cache the hashes for each package.
-  let hash = cachedPackageVersionHashes[moduleName];
-  if (!hash) {
-    const version = getPackageVersion(moduleName);
-    hash = crypto.createHash('sha256').update(version).digest('hex').slice(0, 16);
-    cachedPackageVersionHashes[moduleName] = hash;
-  }
-
-  return `/cacheable_node_modules/${hash}/${assetPath}`;
+  return `${assetsPrefix}/node_modules/${assetPath}`;
```

Comment:
Now that we have a "global" cachebuster, I removed the package specific version hash. The only downside to this is that if some package `X` isn't modified by another package is (and thus the cachebuster changes), we'll still get cache misses for `X`; in the old version, the hash was based on the package version, which meant that we'd only get misses when the package was actually updated. I think this is a reasonable tradeoff for now, and will be essentially a non-issue once we're behind a CDN that can cache this stuff.

Reply:
That was a nice property to enable fine-grained hashing. Would it be much additional code complexity to handle `node_modules` using the per-package cache-busting and only use the "global cache-busting" for `public` and `elements`? I'm thinking that between dependabot runs we normally don't need to invalidate the `node_modules` caches because it's only our `build` code that's changing most of the time. For example, at the moment we'll be forcing clients to re-download all of mathjax (including fonts) every time we change one of our transpiled client files.

Going further down the fine-grained rabbit hole, I almost wonder whether could cache-bust each element separately?

Reply:
I don't think that's an unreasonable thing to do. I'm envisioning routes like these (using the default `/assets` prefix):

- `/assets/node_modules/CACHEBUSTER/bootstrap/...`
- `/assets/public/CACHEBUSTER/colors.css/...`
- `/assets/elements/pl-file-editor/CACHEBUSTER/pl-file-editor.js`
- `/assets/build/foo-CACHEBUSTER.js`

Of course, `CACHEBUSTER` would be scoped to each respective route "type".

If you like this, I can implement it!

Reply:
I implemented the fine-grained approach in 1815fd7. It includes hashes for individual files within `elements` and `public`, and also includes package-version-based hashes for things within `node_modules`.

Reply:
Awesome!